New module: Add module for managing AWS Datapipelines (cloud/amazon/data_pipeline)

s-hertel opened the pull request:
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/amazon/data_pipeline.py

##### ANSIBLE VERSION
```
ansible 2.4.0 (ansible-modules-extras/pull/1547 8b8a039e4c) last updated 2017/03/28 13:22:30 (GMT -400)
  config file =
  configured module search path = Default w/o overrides
```

##### SUMMARY
- Supports create/activate/deactivate and deletion
- Handles idempotent creation by hashing the options (except `objects`) given to the module
- Waits for requested state to be reached, as Botocore doesn't
  have waiters yet for datapipelines

Commit:
New module for managing AWS Datapipelines

* Supports create/activate/deactivate and deletion
* Handles idempotent creation by embeding the version in the
uniqueId field
* Waits for requested state to be reached, as Botocore doesn't
have waiters yet for datapipelines

s-hertel commented:
Migrated from https://github.com/ansible/ansible-modules-extras/pull/1547 by s-hertel (not original author)

wimnat commented on lib/ansible/modules/cloud/amazon/datapipeline.py:
```diff
@@ -0,0 +1,470 @@
+
+DOCUMENTATION = '''
+---
+module: datapipeline
```
I would prefer `data_pipeline` over `datapipeline`

wimnat commented on lib/ansible/modules/cloud/amazon/datapipeline.py:
```diff
@@ -0,0 +1,470 @@
+  state:
+    description:
+      - Requested state of the pipeline
+    choices: ['present', 'absent', 'active', 'deactive']
```
Should be `inactive` rather than `deactive`

s-hertel commented:
Thanks wimnat. When I fix this up I will also make the changes you requested [previously](https://github.com/ansible/ansible-modules-extras/pull/1547).

s-hertel commented on lib/ansible/modules/cloud/amazon/datapipeline.py:
```diff
@@ -0,0 +1,470 @@
+        return False
+
+
+def pipeline_status_timeout(client, dp_id, status, timeout):
```
Unfortunately, there seems to be no built in waiter for datapipelines - get_waiter() returns an empty list. Going to keep this as is.

s-hertel commented on lib/ansible/modules/cloud/amazon/datapipeline.py:
```diff
@@ -0,0 +1,470 @@
+        raise
+
+
+def pipeline_exists_timeout(client, dp_id, timeout):
```
As above, keeping as is since there is no built in waiter.

s-hertel committed:
rename module, fix imports, add tags option, improve exit_json results, fix a couple bugs, add a TODO so I don't forget

fix pep8

allow timeout to be used for pipeline creation

make .format syntax uniform

fix pep8

fix exception handling

allow pipeline to be modified, refactor, add some comments, remove unnecessary imports

pipeline activation may not be in the activated state long

remove datapipeline version option

change a loop to a list comprehension

create idempotence by hashing the options given to the module minus the objects (which can be modified)

small bugfix

s-hertel committed:
data_pipeline unittests

make unittests pep8

fix bug in unittests

mattclay commented:
CI failure in unit tests. 5 failures for python 2.x and 15 for 3.x:

https://app.shippable.com/github/ansible/ansible/runs/18680/2/tests
https://app.shippable.com/github/ansible/ansible/runs/18680/4/console

s-hertel committed:
remove exception handling that serves no purpose

mattclay commented:
CI failure in python 3.x unit tests:

https://app.shippable.com/github/ansible/ansible/runs/18770/4/tests
https://app.shippable.com/github/ansible/ansible/runs/18770/5/tests

All of the failures appear to be related:

```
>       hashed_data = hashlib.md5(json_data).hexdigest()
E       TypeError: Unicode-objects must be encoded before hashing

lib/ansible/modules/cloud/amazon/data_pipeline.py:400: TypeError
```

s-hertel committed:
Fix python3 incompatibilities in datapipeline tests and add placebo fixture maybe_sleep for faster tests

Fix python3 incompatibilities in data_pipeline build_unique_id()

Don't delete a pipeline in diff_pipeline() because it's unexpected

Don't use time.time() because it causes an issue with placebo testing

re-recorded tests

fix pep8 in data_pipeline

Remove disable_rollback from tests

Make sure unique identifier is a string

re-record tests

s-hertel commented:
ready_for_review

ryansb commented on lib/ansible/modules/cloud/amazon/data_pipeline.py:
```diff
@@ -0,0 +1,594 @@
+    default: ''
+  objects:
+    description:
+      - List of pipeline Objects
```
Is there a link specific to the structure of these objects? I like to include as much info about complex types in the docs as possible.

s-hertel replied:
Should I move the link on [line 37](https://github.com/ansible/ansible/pull/22878/files#diff-03217c7dfd53337741c986624bc4e7c9R37) here? I'll also add `suboptions` here to make it clearer what is expected.

ryansb commented on lib/ansible/modules/cloud/amazon/data_pipeline.py:
```diff
@@ -0,0 +1,594 @@
+    required: false
+  values:
+    description:
+      - List of pipeline values
```
Same here

ryansb commented on lib/ansible/modules/cloud/amazon/data_pipeline.py:
```diff
@@ -0,0 +1,594 @@
+    :returns: True or False
+
+    """
+    assert type(status) == list
```
Instead of equality, I think you want to check for `isinstance` here.

ryansb commented on test/units/modules/cloud/amazon/test_data_pipeline.py:
```diff
@@ -0,0 +1,275 @@
+    return m, connection, result['data_pipeline']['pipeline_id']
+
+
+def tear_down_pipeline(connection, module):
```
This sort of work is best done as a pytest fixture. With [fixtures](https://docs.pytest.org/en/latest/fixture.html) you could write it such that one fixture would set up & tear down the whole setup. 

```
@pytest.fixture
def pipeline_fixture(placeboify, objects, tags):
    connection = placeboify.client('datapipeline')
    [insert params stuff and FakeModule]
    _, result = db.create_pipeline(connection, m)
    yield {'module': m, 'connection': connection, 'pipeline_id': result['data_pipeline']['pipeline_id']}
    
    # this part runs after the test finishes, pass or fail
    dp.delete_pipeline({'name': 'ansible-test-create-pipeline', 'state': 'absent'})
```

And that would merge the setup/teardown functions into a fixture that you could use in tests like:

```
test_define_pipeline(pipeline_fixture, maybe_sleep):
    pipeline_fixture['module'].[call a method or what have you](pipeline_fixture['pipeline_id'])
```

s-hertel committed:
improve documentation and add another example

s-hertel committed:
Fix data_pipeline docs

use isinstance instead of type()

fix documentation

s-hertel committed:
use a placebo fixture instead of redundant code in tests

fix tests for PLACEBO_RECORD=false

s-hertel committed:
fix documentation

s-hertel commented:
ready_for_review

ryansb commented on lib/ansible/modules/cloud/amazon/data_pipeline.py:
```diff
@@ -0,0 +1,616 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+#
+# This file is part of Ansible
```
New license block: http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#documentation-block

ryansb commented on lib/ansible/modules/cloud/amazon/data_pipeline.py:
```diff
@@ -0,0 +1,616 @@
+---
+module: data_pipeline
+version_added: "2.4"
+author: Raghu Udiyar <raghusiddarth@gmail.com> (@raags)
```
You can definitely add yourself here as well.

ryansb reviewed (commented):
Minor docs things, otherwise looks good.

s-hertel commented:
@wimnat Would you please re-review this?

s-hertel committed:
Remove use of undefined variable from data_pipeline module and fix license

s-hertel committed:
fix copyright header
